src/udev/udev-event.c

```diff
@@ -346,12 +346,30 @@ size_t udev_event_apply_format(struct udev_event *event,
                         } else if (from[0] == '%') {
                                 /* substitute format char */
                                 unsigned int i;
+                                unsigned int num = 0;
+                                char from1 = from[1];
 
                                 if (from[1] == '%') {
                                         from++;
                                         goto copy;
                                 }
 
+                                /* extract possible format length and move from behind it*/
+                                if (isdigit(from1)) {
+                                        char *tail;
+
+                                        from++;
+                                        num = (int) strtoul(from, &tail, 10);
+                                        if (num > 0) {
+                                                log_debug("format length='%i'", num);
+                                                from = strjoina("%", tail);
+                                                l = num+1;
```

Comment:
num+1 > size should be handeled by the backend function.
The function strpcpy (src/basic/strxcpyx.c) that is called from subst_format_var will only copy up to the length of the source string:

size_t strpcpy(char **dest, size_t size, const char *src) {
[...]
        len = strlen(src);
        if (len >= size) {
                if (size > 1)
                        *dest = mempcpy(*dest, src, size-1);
                size = 0;
        } else {
                if (len > 0) {
                        *dest = mempcpy(*dest, src, len);
                        size -= len;
                }
        }